pkg/compactor/compactor_test.go

```diff
@@ -1364,6 +1364,129 @@ func TestMultitenantCompactor_ShouldSkipCompactionForJobsNoMoreOwnedAfterPlannin
 	))
 }
 
+func TestMultitenantCompactor_ShouldSkipCompactionForJobsWithFirstLevelCompactionBlocksAndWaitPeriodNotElapsed(t *testing.T) {
+	t.Parallel()
+
+	const waitPeriod = 10 * time.Minute
+
+	// Mock two tenants, each with the same exact overlapping blocks. However,
+	// the 2nd tenant has uploaded the blocks less than "wait period" ago.
+	bucketClient := &bucket.ClientMock{}
+	bucketClient.MockIter("", []string{"user-1", "user-2"}, nil)
+
+	for _, userID := range []string{"user-1", "user-2"} {
+		lastModified := time.Now().Add(-20 * time.Minute)
+		if userID == "user-2" {
+			lastModified = time.Now().Add(-5 * time.Minute)
+		}
+
+		bucketClient.MockExists(path.Join(userID, mimir_tsdb.TenantDeletionMarkPath), false, nil)
+		bucketClient.MockIter(userID+"/", []string{path.Join(userID, "/01DTVP434PA9VFXSW2JK000001"), path.Join(userID, "/01DTVP434PA9VFXSW2JK000002")}, nil)
+		bucketClient.MockIter(userID+"/markers/", nil, nil)
+		bucketClient.MockGetAndLastModified(path.Join(userID, "01DTVP434PA9VFXSW2JK000001/meta.json"), mockBlockMetaJSONWithTimeRange("01DTVP434PA9VFXSW2JK000001", 1574776800000, 1574784000000), lastModified, nil)
+		bucketClient.MockGet(path.Join(userID, "01DTVP434PA9VFXSW2JK000001/deletion-mark.json"), "", nil)
+		bucketClient.MockGet(path.Join(userID, "01DTVP434PA9VFXSW2JK000001/no-compact-mark.json"), "", nil)
+		bucketClient.MockGetAndLastModified(path.Join(userID, "01DTVP434PA9VFXSW2JK000002/meta.json"), mockBlockMetaJSONWithTimeRange("01DTVP434PA9VFXSW2JK000002", 1574776800000, 1574784000000), lastModified, nil)
+		bucketClient.MockGet(path.Join(userID, "01DTVP434PA9VFXSW2JK000002/deletion-mark.json"), "", nil)
+		bucketClient.MockGet(path.Join(userID, "01DTVP434PA9VFXSW2JK000002/no-compact-mark.json"), "", nil)
+		bucketClient.MockGet(path.Join(userID, "bucket-index.json.gz"), "", nil)
+		bucketClient.MockUpload(path.Join(userID, "bucket-index.json.gz"), nil)
+	}
+
+	cfg := prepareConfig(t)
+	cfg.CompactionWaitPeriod = waitPeriod
+	c, _, tsdbPlanner, logs, registry := prepare(t, cfg, bucketClient)
+
+	// Mock the planner as if there's no compaction to do, in order to simplify tests.
+	tsdbPlanner.On("Plan", mock.Anything, mock.Anything).Return([]*metadata.Meta{}, nil)
+
+	require.NoError(t, services.StartAndAwaitRunning(context.Background(), c))
+
+	// Compactor doesn't wait for blocks cleaner to finish, but our test checks for cleaner metrics.
+	require.NoError(t, c.blocksCleaner.AwaitRunning(context.Background()))
+
+	// Wait until a run has completed.
+	test.Poll(t, 5*time.Second, 1.0, func() interface{} {
+		return prom_testutil.ToFloat64(c.compactionRunsCompleted)
+	})
+
+	require.NoError(t, services.StopAndAwaitTerminated(context.Background(), c))
+
+	// We expect only 1 compaction job has been expected, while the 2nd has been skipped.
+	tsdbPlanner.AssertNumberOfCalls(t, "Plan", 1)
+
+	assert.ElementsMatch(t, []string{
```

Comment:
I know this isn't the first test to do this, but I really don't like asserting on side effects of an operation (`info` log statements) instead of the results (some sort of "compaction result" object?).

Reply:
I see your point. I've simplified it in https://github.com/grafana/mimir/pull/4401/commits/6e8f723a12df1fdf6054d1e4faf10372898570c6. I'm still checking logs, but asserting only on 1 log line, which is what we care about in this test.